rest/route/middleware.go

```diff
@@ -287,3 +293,89 @@ func (m *CommitQueueItemOwnerMiddleware) ServeHTTP(rw http.ResponseWriter, r *ht
 
 	next(rw, r)
 }
+
+func RequiresProjectPermission(permission string, level int) gimlet.Middleware {
+	if !evergreen.AclCheckingIsEnabled {
+		return &noopMiddleware{}
+	}
+
+	opts := gimlet.RequiresPermissionMiddlewareOpts{
+		RM:            evergreen.GetEnvironment().RoleManager(),
+		PermissionKey: permission,
+		ResourceType:  "project",
+		RequiredLevel: level,
+		ResourceFunc:  urlVarsToScopes,
+	}
+	return gimlet.RequiresPermission(opts)
+}
+
+type noopMiddleware struct{}
+
+func (n *noopMiddleware) ServeHTTP(rw http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
+	next(rw, r)
+}
+
+func urlVarsToScopes(r *http.Request) string {
+	vars := gimlet.GetVars(r)
+	query := r.URL.Query()
+
+	projectID := util.CoalesceStrings(append(query["project_id"], query["projectId"]...), vars["project_id"], vars["projectId"])
+	if projectID != "" {
+		return projectID
+	}
+
+	versionID := util.CoalesceStrings(append(query["version_id"], query["versionId"]...), vars["version_id"], vars["versionId"])
+	if versionID != "" {
+		v, err := model.VersionFindOne(model.VersionById(versionID).Project(bson.M{model.VersionIdentifierKey: 1}))
```

Comment:
I feel pretty weird about calling model code here rather than using the service context

Reply:
Abstracted some of the query into the models package